server/container_create.go

```diff
@@ -359,6 +359,19 @@ func (s *Server) CreateContainer(ctx context.Context, req *types.CreateContainer
 		}, nil
 	}
 
+	systemCtx, err := s.contextForNamespace(req.SandboxConfig.Metadata.Namespace)
+	if err != nil {
+		return nil, fmt.Errorf("get context for namespace: %w", err)
+	}
+
+	if err := s.StorageImageServer().IsRunningImageAllowed(ctx, &systemCtx, req.Config.Image.UserSpecifiedImage); err != nil {
```

Comment:
Does that mean we don't need the `req.Config.Image.UserSpecifiedImage` at all? 

If we switch that to `req.Config.Image.Image`, then we need additional logic for those two tests:

https://github.com/cri-o/cri-o/blob/efd1654b9994ddbc1fd9f563a817d98b128dfd89/test/policy.bats#L144-L175


Otherwise we will get the error:

```
# time="2023-07-28 09:43:28.733989619+02:00" level=debug msg="Response error: parsing user specified image: invalid repository name (2f1bf9c599183e0ef813e9b76a8cdd17c9b2cb224c94b605c178dcef8d3a54e3), cannot specify 64-byte hexadecimal strings" file="otel-collector/interceptors.go:71" id=31cde433-4fda-4373-a473-6f438172e1d7 name=/runtime.v1.RuntimeService/CreateContainer
```

If we have an image ID in `req.Config.Image.Image`, should we then fallback to `req.Config.Image.UserSpecifiedImage`?

Reply:
> Does that mean we don't need the `req.Config.Image.UserSpecifiedImage` at all?

No?! Policy verification has three main inputs:
- what is the users’ intent / instruction
- what are the rules for deciding whether something matches the users’ intent
- what is being verified against that intent

For us, the three are
- `UserSpecifiedImage`
- `policy.json`
- A single `*storage.Image`.

---

> If we have an image ID in `req.Config.Image.Image`, should we then fallback to `req.Config.Image.UserSpecifiedImage`?

What would that _mean_ in terms of the three inputs discussed above?

---

The `PullImage`+`CreateContainer` pair needs to be considered together. IIRC the situation is one of:
- a fresh pull
  - Pull the image, with policy enforcement included
  - Get a manifest digest from `copy.Image`, construct _the right_ `RepoDigest` value, return that in `PullImageResponse`
  - Receive exactly that `RepoDigest` value in `Config.Image.Image`
  - Use it to look up a `*storage.Image` (`NewStoreReference(store, repoDigest, "")`).
  - Verify signatures against policy and `UserSpecifiedImage`
- a pre-existing image reuse short path:
  - Find a candidate _commit_ to an `*Image.ID`
  - Verify a registry match
  - Verify that image’s signature on the pull short path (`NewStoreReference(store, nil, imageID)` or `NewStoreReference(store, RepoDigest, imageID)` if we have a repoDigest input — or, well, ideally, get a `*storage.Image` directly from the local lookup without going through an `*Image` -> ID -> `*Image` roundtrip), against the image we were requested to pull
  - Return _the right_ RepoDigest if known, or an image ID (if not-pulling by tag, IIRC) in `PullImageResponse`
  - If we received an ID in `Config.Image.Image`, use it to look up an image, possibly combined with UserSpecifiedImage (same `NewStoreReference` calls)
  - Verify signatures against policy and `UserSpecifiedImage`

Reply:
> Verify a registry match

Hum, that could potentially be used to obtain the right manifest digest if users request a pull by tag. But, in principle, that does matter because…

---

… there is a third situation, per https://github.com/cri-o/cri-o/pull/7046#discussion_r1228433790 : Where there never is an `ImagePull`, and we get a `CreateContainer` with an image ID, based on Kubelet calling `ImageStatus`. So `CreateContainer` must be ready to accept repo-less digest IDs either way.